pallets/moonbeam-lazy-migrations/src/lib.rs

```diff
@@ -103,33 +104,93 @@ pub mod pallet {
 				Error::<T>::AllStorageEntriesHaveBeenRemoved
 			);
 
-			let hashed_prefix = sp_io::hashing::twox_128("LocalAssets".as_bytes());
+			let mut allowed_removals = limit;
 
-			let keys_removed = match sp_io::storage::clear_prefix(&hashed_prefix, Some(limit)) {
-				sp_io::KillStorageResult::AllRemoved(value) => {
-					LocalAssetsMigrationCompleted::<T>::set(true);
-					weight
-						.saturating_accrue(<T as frame_system::Config>::DbWeight::get().writes(1));
-					value
+			const PALLET_PREFIX: &'static str = "LocalAssets";
+
+			struct LocalAssetsStorageAsset;
+			impl frame_support::traits::StorageInstance for LocalAssetsStorageAsset {
+				const STORAGE_PREFIX: &'static str = "Asset";
+				fn pallet_prefix() -> &'static str {
+					PALLET_PREFIX
+				}
+			}
+			struct LocalAssetsStorageApprovals;
+			impl frame_support::traits::StorageInstance for LocalAssetsStorageApprovals {
+				const STORAGE_PREFIX: &'static str = "Approvals";
+				fn pallet_prefix() -> &'static str {
+					PALLET_PREFIX
 				}
-				sp_io::KillStorageResult::SomeRemaining(value) => value,
-			} as u64;
-
-			log::info!("Removed {} keys 🧹", keys_removed);
-
-			Ok(Some(
-				weight
-					.saturating_add(Weight::from_parts(
-						0,
-						INTERMEDIATES_NODES_SIZE
-							+ MAX_LOCAL_ASSETS_STORAGE_ENTRY_SIZE * keys_removed,
-					))
-					.saturating_add(
-						<T as frame_system::Config>::DbWeight::get()
-							.reads_writes(keys_removed, keys_removed),
+			}
+
+			/// Data concerning an approval.
+			/// Duplicated the type from pallet_assets (The original type is private)
+			#[derive(
+				Clone, Encode, Decode, Eq, PartialEq, RuntimeDebug, Default, MaxEncodedLen, TypeInfo,
+			)]
+			pub struct Approval<Balance, DepositBalance> {
+				/// The amount of funds approved for the balance transfer from the owner to some delegated
+				/// target.
+				pub(super) amount: Balance,
+				/// The amount reserved on the owner's account to hold this item in storage.
+				pub(super) deposit: DepositBalance,
+			}
+
+			let asset_id_iter = frame_support::storage::types::StorageMap::<
+				LocalAssetsStorageAsset,
+				Blake2_128Concat,
+				u128,
+				// It is fine to add a dummy `Value` type
+				// The value is not going to be decoded, since we only care about the keys)
+				(),
+			>::iter_keys();
+
+			for asset_id in asset_id_iter {
+				let approvals_iter = frame_support::storage::types::StorageNMap::<
+					LocalAssetsStorageApprovals,
+					(
+						NMapKey<Blake2_128Concat, u128>,         // asset id
+						NMapKey<Blake2_128Concat, T::AccountId>, // owner
+						NMapKey<Blake2_128Concat, T::AccountId>, // delegate
 					),
-			)
-			.into())
+					Approval<T::Balance, T::Balance>,
+				>::drain_prefix((asset_id,));
+				for ((owner, _), approval) in approvals_iter {
```

Comment:
We should account proof size for each Approvals entry write

Reply:
This is already done by the `limit` argument, we compute the worst case cost with this threshold.

Reply:
@RomarQ I speak about the proof size and not the ref time, keep in mind that accounting db reads/writes doesn't account proof size